src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,457 @@
+    @Override
+    public void visitToken(DetailAST ast) {
+        final int type = ast.getType();
+        if (type == TokenTypes.DOT) {
+            visitDotToken(ast);
+        }
+        else if (type == TokenTypes.VARIABLE_DEF) {
+            visitVariableDefToken(ast);
+        }
+        else if (type == TokenTypes.IDENT) {
+            visitIdentToken(ast);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast, SCOPES)) {
+            while (!variables.isEmpty() && variables.peek().getScope() == ast) {
+                final VariableDesc variableDesc = variables.pop();
+                if (!variableDesc.isUsed()) {
+                    final DetailAST typeAst = variableDesc.getTypeAst();
+                    log(typeAst, MSG_UNUSED_LOCAL_VARIABLE, variableDesc.name);
+                }
+            }
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#DOT}.
+     *
+     * @param dotAst dotAst
+     */
+    private void visitDotToken(DetailAST dotAst) {
+        if (shouldCheckIdentTokenNestedUnderDot(dotAst)) {
+            checkIdentifierAst(dotAst.findFirstToken(TokenTypes.IDENT));
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#VARIABLE_DEF}.
+     *
+     * @param varDefAst varDefAst
+     */
+    private void visitVariableDefToken(DetailAST varDefAst) {
+        if (varDefAst.getParent().getType() != TokenTypes.OBJBLOCK) {
+            final DetailAST ident = varDefAst.findFirstToken(TokenTypes.IDENT);
+            final VariableDesc desc = new VariableDesc(ident.getText(),
+                    varDefAst.findFirstToken(TokenTypes.TYPE), findScopeOfVariable(varDefAst));
+            variables.push(desc);
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#IDENT}.
+     *
+     * @param identAst identAst
+     */
+    private void visitIdentToken(DetailAST identAst) {
+        final DetailAST parentAst = identAst.getParent();
+        if (!TokenUtil.isOfType(parentAst, UNACCEPTABLE_PARENT_OF_IDENT)
+                && shouldCheckIdentWithMethodRefParent(identAst)) {
+            checkIdentifierAst(identAst);
+        }
+    }
+
+    /**
+     * Whether an ident with parent node of type {@link TokenTypes#METHOD_REF}
+     * should be checked or not.
+     *
+     * @param identAst identAst
+     * @return true if an ident with parent node of type {@link TokenTypes#METHOD_REF}
+     *         should be checked or if the parent type is not {@link TokenTypes#METHOD_REF}
+     *         else false
+     */
+    public static boolean shouldCheckIdentWithMethodRefParent(DetailAST identAst) {
+        final DetailAST parent = identAst.getParent();
+        boolean result = true;
+        if (parent.getType() == TokenTypes.METHOD_REF) {
+            result = parent.getFirstChild() == identAst
+                    && parent.getLastChild().getType() != TokenTypes.LITERAL_NEW;
+        }
+        return result;
+    }
```

Comment:
Can refactor this into-
```
public static boolean shouldCheckIdentWithMethodRefParent(DetailAST identAst) {
        final DetailAST parent = identAst.getParent();
        return parent.getType() != TokenTypes.METHOD_REF || parent.getFirstChild() == identAst
                && parent.getLastChild().getType() != TokenTypes.LITERAL_NEW;
}
```
But readability is better before refactoring, after refactoring it becomes difficult to read and the condition is not that clear.